zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+/// and is of length up to `n`. The range is truncated if necessary to end at
+/// the maximum valid address index, `i32::MAX`.
+///
+/// Precondition: `i >= -1 and n > 0`
+pub(crate) fn range_after(i: i32, n: i32) -> RangeInclusive<u32> {
+    assert!(i >= -1);
+    assert!(n > 0);
+    let first = u32::try_from(i64::from(i) + 1).unwrap();
+    let last = u32::try_from(i.saturating_add(n)).unwrap();
+    first..=last
+}
+
+/// Returns the ephemeral transparent IVK for a given account ID.
+pub(crate) fn get_ephemeral_ivk<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+) -> Result<EphemeralIvk, SqliteClientError> {
+    Ok(get_account(conn, params, account_id)?
+        .ok_or(SqliteClientError::AccountUnknown)?
+        .ufvk()
+        .and_then(|ufvk| ufvk.transparent())
+        .ok_or(SqliteClientError::UnknownZip32Derivation)?
+        .derive_ephemeral_ivk()?)
+}
+
+/// Returns a vector with all ephemeral transparent addresses potentially belonging to this wallet.
+/// If `for_detection` is true, this includes addresses for an additional GAP_LIMIT indices.
+pub(crate) fn get_reserved_ephemeral_addresses<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+    for_detection: bool,
+) -> Result<HashMap<TransparentAddress, Option<TransparentAddressMetadata>>, SqliteClientError> {
+    let mut stmt = conn.prepare(
+        "SELECT address, address_index FROM ephemeral_addresses WHERE account_id = :account ORDER BY address_index",
+    )?;
+    let mut rows = stmt.query(named_params! { ":account": account_id.0 })?;
+
+    let mut result = HashMap::new();
+    let mut first_unused_index: Option<i32> = Some(0);
+
+    while let Some(row) = rows.next()? {
+        let addr_str: String = row.get(0)?;
+        let raw_index: u32 = row.get(1)?;
+        first_unused_index = i32::try_from(raw_index)
+            .map_err(|e| SqliteClientError::CorruptedData(e.to_string()))?
+            .checked_add(1);
+        let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+        let address = TransparentAddress::decode(params, &addr_str)?;
+        result.insert(address, Some(metadata(address_index)));
+    }
+
+    if for_detection {
+        if let Some(first) = first_unused_index {
+            let ephemeral_ivk = get_ephemeral_ivk(conn, params, account_id)?;
+
+            for raw_index in range_after(first, GAP_LIMIT) {
+                let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+                let address = ephemeral_ivk.derive_ephemeral_address(address_index)?;
+                result.insert(address, Some(metadata(address_index)));
+            }
+        }
+    }
+    Ok(result)
+}
+
+/// Returns a vector with the next `n` previously unreserved ephemeral addresses for
+/// the given account.
+///
+/// Precondition: `n < 0x80000000`
```

Comment:
In Rust APIs, this is generally encoded in documentation as:
```suggestion
/// # Panics
///
/// Panics if `n >= 0x80000000`.
```